docs/design/1-overview.md

```diff
@@ -0,0 +1,520 @@
+    And Mark should not receive any error description
+
+  Scenario: Query a job (success)
+    Given Mark has auth certificates in place
+    When Mark requests to query a job
+    And there are no internal system errors
+    And the job exists
+    Then Mark should receive a success response
+    And Mark should receive the job information
+    But Mark should not receive output of the job
+  
+  Scenario: Query a job (failure)
+    Given Mark has auth certificates in place
+    When Mark requests to query a job
+    And the job exists
+    Then Mark should receive an unauthorised response
+    And Mark should not reveive any job information
+    And Mark should not receive any error description
+
+  Scenario: Query a job (not_exist)
+    Given Mark has auth certificates in place
+    When Mark requests to query a job
+    And there are no internal system errors
+    And the job does not exist
+    Then Mark should receive an unauthorised response
+    And Mark should not be able to determine if the job existed.
+
+  Scenario: Query a job (stream_output)
+    Given Mark has auth certificates in place
+    When Mark requests to query a job with output stream
+    And there are no internal system errors
+    And the job exists
+    Then Mark should receive a success response
+    And Mark should receive an output stream for the job
+    But the stream should include the entire output history
+  
+```
+
+---
+
+# 🏗️ Architectural Overview
+
+## 🖼️ High-Level Diagram
+
+![1-job-worker-hex-arch.png](images/1-job-worker-hex-arch.png)
+
+## 🛠️ Key Components & Responsibilities
+Describe major modules, services, or layers in the system.
+
+System will be a layered architecture. Domain logic will interact with an abstraction (service) of the lowest level worker package, ensuring the system can be extended.
+
+### Worker Package (Library)
+The worker package will use the following interface structure (subject to refinement);
+
+```go
+const cgroupRoot = "/sys/fs/cgroup" // Adjust based on cgroup version
+
+// Command represents a job request with resource limits.
+type Command struct {
+	Name     string            // Command name (e.g., "ping") 
+	Args     []string          // Command arguments (e.g., ["-c", "4", "google.com"])
+	ResourceLimits CgroupConfig
+}
+
+// CgroupConfig defines resource limits for a job using cgroup v2.
+type CgroupConfig struct {
+        CPUQuota     string // Max CPU usage (e.g., "50000 100000" for 50% CPU)
+        MemoryMax    int64  // Max memory limit in bytes (e.g., 500MB = 500*1024*1024)
+        MemorySwap   int64  // Max swap usage in bytes (0 = no swap, -1 = unlimited)
+        DiskReadBPS  int64  // Max disk read speed in bytes per second (e.g., 5MB/s = 5*1024*1024)
+        DiskWriteBPS int64  // Max disk write speed in bytes per second (e.g., 2MB/s = 2*1024*1024)
+}
+
+// Job represents a running process with cgroup tracking.
+type Job struct {
+	ID     string             // Job identifier
```

Comment:
How will you generate the job ID?

Reply:
It will be a UUID: https://pkg.go.dev/github.com/google/uuid